### PR TITLE
ISPN-5857 ISPN-5858 ISPN-5859 ISPN-5860 HotRod performance improvements for replicated caches

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.jboss.org/browse/ISPN-5857
https://issues.jboss.org/browse/ISPN-5858
https://issues.jboss.org/browse/ISPN-5859
https://issues.jboss.org/browse/ISPN-5860

Just for fun, are the flamegraphs for the local-only test before and after:

* https://dl.dropboxusercontent.com/u/138257/local_master.svg
* https://dl.dropboxusercontent.com/u/138257/local_after_fix.svg

### CODE REVIEW

**Comment by gustavonalle** [comment]
There are some hot rod 1.x test failures

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/retry/ReplicationHitsTest.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.remoting.rpc.RpcManagerImpl;
+import org.testng.annotations.Test;
+
+import static org.testng.AssertJUnit.assertEquals;
+
+/**
+ * @author Dan Berindei
+ * @since 6.2
```
Comment (gustavonalle) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/retry/ReplicationHitsTest.java]: 6.2?

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/Encoder2x.scala [hunk 2]
```diff
@@ -51,7 +49,21 @@ object Encoder2x extends AbstractVersionedEncoder with Constants with Log {
...
    override def writeHeader(r: Response, buf: ByteBuf, addressCache: AddressCache, server: HotRodServer): Unit = {
-      val newTopology = getTopologyResponse(r, addressCache, server)
+      // Sometimes an error happens before we have added the cache to the knownCaches/knownCacheConfigurations map
+      // If that happens, we pretend the cache is LOCAL and we skip the topology update
+      val cr = server.getCacheRegistry(r.cacheName)
+      val configuration = server.getCacheConfiguration(r.cacheName)
+      val cacheMode = configuration match {
+         case null => CacheMode.LOCAL
+         case _ => configuration.clustering().cacheMode()
+      }
```
Comment (gustavonalle) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/Encoder2x.scala]: Nitpick: 
```
val cacheMode = Option(configuration).map(_.clustering.cacheMode).getOrElse(CacheMode.LOCAL)
```

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/Encoder2x.scala [hunk 3]
```diff
@@ -51,7 +49,21 @@ object Encoder2x extends AbstractVersionedEncoder with Constants with Log {
...
+      val cacheTopology = cacheMode.isClustered match {
+         case false => null
+         case _ => cr.getStateTransferManager.getCacheTopology
+      }
```
Comment (gustavonalle) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/Encoder2x.scala]: Nitpick: 
``` val cacheTopology = if(cacheMode.isClustered) cr.getStateTransferManager.getCacheTopology else null```

Comment (danberindei) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/Encoder2x.scala]: Nice, I was looking for a replacement for the `?:` operator in Scala but I missed `else`.

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/Encoder2x.scala (hunk 2 above)
Comment (danberindei) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/Encoder2x.scala]: I feel like the extra `Option` and methods would obscure what I'm trying to do...

**Comment by danberindei** [comment]
Updated, the failures should be fixed now.

**Comment by gustavonalle** [comment]
pulling

**Comment by gustavonalle** [comment]
Integrated, thanks @danberindei

**Pull request closed without merging**